### ISSUE TITLE
userInfo Erased Before Reading from Cache

### ISSUE THREAD

**Issue by Sam-Spencer** [kean/Nuke#539]
Setting any `userInfo` on an `ImageRequest` and attempting to retrieve it from an `ImageResponse` always results in an empty `userInfo` object.

It appears that during Image Decoding (at least in `ImageDecoders+Default.swift`) the `userInfo` information gets discarded, particularly in this function:

```swift
private func _decode(_ data: Data) -> ImageContainer? {
    ...
}
```

**Comment by kean** [comment]
`userInfo` in `ImageRequest` is for passing additional parameters of the request. You can access this `userInfo` in any of your custom data loaders, decoders, etc.

`userInfo` in `ImageResponse` is for sending additional metadata about the response. For example, in a progressive JPEG decoder it sends a number of decoded scans:

```swift
return ImageContainer(image: image, type: .jpeg, isPreview: true, userInfo: [.scanNumberKey: numberOfScans])
```

**Comment by kean** [comment]
> If this behavior is intentional, I'm curious why?

The only scenario where I ever needed it was to pass additional parameters to custom parts of the pipeline, e.g. decoders.

> TL;DR: It would be super nice if userInfo was propagated through from the ImageRequest to the ImageResponse

Yeah, it seems like a nice improvement.

> is different from all other system

Do you have some examples? I remembered `userInfo` in `NSNotification`, but it's just for attaching additional information to the notification object. It's the same with `ImageRequest` and `ImageResponse`. The user can attach additional information to `ImageRequest`, and the user can attach additional information to `ImageResponse`. In that sense, I think it works the same way as `NSNotification`. But yes, it can be confusing because the property is named the same for both `ImageRequest` and `ImageResponse`.

**Comment by Sam-Spencer** [comment]
> Yeah, it seems like a nice improvement.

Would you mind if I submit a PR for this -- assuming the change is purely additive?

> Do you have some examples?

I might be remembering this type of behavior from those long-ago Objective-C days 😅 , but in Swift here are a few examples that preserve what's passed in and spit it back out later:
 - [NSUserActivity](https://developer.apple.com/documentation/foundation/nsuseractivity/1411706-userinfo) for handoff, continuity, etc.   
 - [UNNotification](https://developer.apple.com/documentation/usernotifications/unnotificationcontent/1649869-userinfo) local notifications  
 - [ClockKit](https://developer.apple.com/documentation/clockkit/clkcomplication/3612131-userinfo) complications   
 - [NSEntityMapping](https://developer.apple.com/documentation/coredata/nsentitymapping/1443184-userinfo) for CoreData migrations  
 - `Error` objects generally leave userInfo dictionaries untouched (although this is probably a poor example because once the error is generated its not usually modified)

**Comment by kean** [comment]
I think the main difference is that in your examples `userInfo` is part of the same type. `ImageRequest` and `ImageResponse` are unrelated types and `userInfo` serves different purposes. You can think of them as `ImageRequest.additionalParameters` and `ImageResponse.metadata` or something like that. At least that's the way it was designed.

**Comment by kean** [comment]
> Would you mind if I submit a PR for this -- assuming the change is purely additive?

Can you check if there are other ways to implement the scenarios you are working on? Since you are calling the API, you know the context and presumably should be able to access it wherever you handle the response.

**Comment by kean** [comment]
I've just had another idea. Maybe the response should contain the original request? I think that's how `URLSession` and some popular third-party frameworks do it. Merging `userInfo` from one type with another unrelated type doesn't seem right.

**Comment by valosip** [comment]
> I've just had another idea. Maybe the response should contain the original request? I think that's how `URLSession` and some popular third-party frameworks do it. Merging `userInfo` from one type with another unrelated type doesn't seem right.

Thats a valid solution for the problem at hand as long as the original request is returned for both a new & cached response!

**Comment by Sam-Spencer** [comment]
👍 Agreed! I see your point about not wanting to merge the `userInfo`. Having the response contain the original request would work great though! It already returns it with remote fetches, just need to pipe it through for local fetching.

**Comment by kean** [comment]
This feature is available in [Nuke 11](https://github.com/kean/Nuke/pull/547) beta.

**Comment by breath103** [comment]
Hi, i'm actually in scenario where you need to preserve some sort of metadata that is derived from response
ex) Extracting average color of loaded image. since that's fairly heavy computation, I'd like to cache the color as part of response. 
such as https://medium.com/swlh/swiftui-read-the-average-color-of-an-image-c736adb43000

@kean can you think of any easy way to implement this on Nuke? 
thanks for the wonderful library BTW :)
